### ISSUE TITLE
addCleanup should run even if setUp fails.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2899]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2899|
|Type|defect|
|Created|2007-11-14 15:47:07Z|
|Branch|https://github.com/twisted/twisted/tree/addCleanup-2899|

Currently, Trial's `TestCase` provides no way to add tear-down code that runs if setUp fails (pyunit shares this bug). `addCleanup`, being a new method, should add cleanup functions that run even if `tearDown` does not.

<details><summary>Searchable metadata</summary>

```
trac-id__2899 2899
type__defect defect
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_addCleanup_2899 branches/addCleanup-2899
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1195055227000000 1195055227000000
changetime__1195076895000000 1195076895000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [21763]) Branching to 'addCleanup-2899'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @jml|
|-|-|

Thank you pyflakes:
```
twisted/trial/test/test_tests.py:704: undefined name 'SkipTest'
```

Otherwise, that looks good :).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set status to `closed`|
|-|-|

(In [21787]) Make cleanup methods added with addCleanup run even if setUp fails.

 * Author: jml
 * Reviewer: therve
 * Fixes [#2899](https://github.com/twisted/twisted/issues/2899) 

Previously, addCleanup cleanup methods ran only if setUp succeeded.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
